### PR TITLE
WIP Multiple limited-scope federations should be deployed

### LINKED ISSUE #557: Install multiple namespaced fedeartion is failed
**What happened**:
1. When second namespaced federation is deployed by deployment script. The operation is failed.

```
$ FEDERATION_NAMESPACE=foo NAMESPACED=y ./scripts/deploy-federation.sh happur/fedv2:test cluster2
...
$ # done successfully

$ FEDERATION_NAMESPACE=dup NAMESPACED=y ./scripts/deploy-federation.sh happur/fedv2:test cluster2
...
F0122 22:15:08.855409    2580 enable.go:117] error: Error creating CRD "federatedclusterroles.primitives.federation.k8s.io": customresourcedefinitions.apiextensions.k8s.io "federatedclusterroles.primitives.federation.k8s.io" already exists
```

2. When skip federate step in the script and manually join cluster. It still failed in
```
I0123 01:25:33.795001    5923 join.go:630] Could not create health check cluster role for service account: cluster2-cluster1 in joining cluster: cluster2 due to: clusterroles.rbac.authorization.k8s.io "federation-controller-manager:healthcheck-cluster2-cluster1" already exists
```
The reason is that our health check uses cluster role.

**What you expected to happen**:
Namespaced federation shold deployed for multiple instances.

**How to reproduce it (as minimally and precisely as possible)**:
see above

**Anything else we need to know?**:
N/A

**Environment**:
- Kubernetes version (use `kubectl version`)
- Federation version
- Install method (helm vs script)
both of them have this issue

- Install configuration (namespaced vs cluster-scoped)
namespaced

- Others

/kind bug

### PR DESCRIPTION
**Pull request by xunpan**
Install multiple namespaced fedeartion is failed. It need to be fixed to make feature 

- [x] multiple namespaced federation deployment
  - [x] kubefedv2 should work correctly with global CRD Federated* and ClusterRole/ClusterRoleBindings
  - [x] scripts should work for deploying and deleting limited-scope federations
- [x] helm chart support to install multiple namespaced federation deployment
- [x] controller log issues
- [ ] more test

Fixes #557

### CODE REVIEW

File: pkg/controller/federatedtypeconfig/controller.go [hunk 1]
```diff
@@ -160,7 +160,7 @@ func (c *Controller) reconcile(qualifiedName util.QualifiedName) util.Reconcilia
 
 	limitedScope := c.controllerConfig.TargetNamespace != metav1.NamespaceAll
 	if limitedScope && syncEnabled && !typeConfig.GetNamespaced() {
-		glog.Infof("Skipping start of sync & status controller for cluster-scoped resource %q.  It is not required for a namespaced federation control plane.", typeConfig.GetFederatedType().Kind)
+		glog.V(3).Infof("Skipping start of sync & status controller for cluster-scoped resource %q.  It is not required for a namespaced federation control plane.", typeConfig.GetFederatedType().Kind)
```
Comment (marun) [inline pkg/controller/federatedtypeconfig/controller.go]: Maybe this message should only be omitted once per kind?

Comment (xunpan) [reply pkg/controller/federatedtypeconfig/controller.go]: I make log level same as 
```
glog.V(3).Infof("Running reconcile FederatedTypeConfig for %q", key)
```
in front of reconcile.

So, when it tells concile, it shows skipping message to cluster-scoped resource. I think we do not need to make more logic for this message.

**Comment by xunpan** [comment]
`enable` should be changed.
E.g. 
federation 1 (namespace 1) enables `MyDepoy` CRD as `FederatedMyDeploy`
In federation 2 (namespace 2), you want to enable `MyDeply` in this federation namespace. It is expected to ignore CRD `FederatedMyDeploy` if it exists and creates a `federatedtypeconfig` for `FederatedMyDeploy` in namespace 2.

**Comment by marun** [comment]
@xunpan How does that help hive though?  enable isn't invoked by hive, it applies pre-generated yaml.

**Comment by xunpan** [comment]
> @xunpan How does that help hive though? enable isn't invoked by hive, it applies pre-generated yaml.

Did you mean openshift/hive? I'm not sure how hive use it. If it use a pre-defined federatedtypeconf resource for a namespaced federation. That's fine. 
But I think `enable` is a general feature to make basic federation case work.

**Comment by marun** [comment]
@xunpan Sorry, I meant *helm*, not *hive*.  How does this PR help helm deploy more than one federation control plane per cluster?

**Comment by xunpan** [comment]
For helm chart:
Installation:
1. make all global  resource created on demand. 
For `CRD`s The method in https://github.com/kubernetes-sigs/federation-v2/issues/557#issuecomment-457160128 is used. If `CRD`s exists, we just skip the `CRD` installation.
For ClusterRoleBinding, e.g.  `federation-controller-manager:healthcheck-cluster1-cluster1`, it is updated with the `SA` created in new namespace to access remote cluster.

2. merge all federated resource `CRD`s into a single file. Then, the above method works for all default configured federated resources as a whole.

Uninstallation:
1. delete namespaced resources
2. delete namespaced `SA` references in `ClusterRoleBinding`, e.g.  by `unjoin`

**Comment by marun** [comment]
@xunpan To my point, the change to `kubefed` has no effect on installation with helm.  Please try to limit the scope of work for any given PR as much as possible.

As to the health check role, are you suggesting that the role binding be shared across control planes?  I think it might make sense to share a role, but why wouldn't each control plane have its own role binding?

**Comment by xunpan** [comment]
@marun I make 2 main commits here.
commit 1: make related command work and script to deploy/delete federation work
commit 2: support helm chart namespaced installation
Did you mean make 2 PRs? 

For health check, I shared clusterrole because all namespaced federation should have same access needs to remote cluster. Yes, we can make each one has their own clusterrole. Any consideration to not share?

**Comment by marun** [comment]
@xunpan Your previous message referred to `ClusterRoleBinding`.  Your most recent message referred to `ClusterRole`.  Which one will be shared?

This PR is changing thousands of lines.  Please provide a clean separation between generated and non-generated changes to simplify review.  And where you can submit a simple cohesive change as a separate PR, please do so.  Reviewing is a hard job even without clean PRs.

**Comment by xunpan** [comment]
`ClusterRoleBinding` binds to `ClusterRole` for permission. It binds to subject to guarantees the permission. 

So, for sharing `ClusterRole`, I mean namespaced `SA` will attached to subject of `ClusterRoleBinding` for sharing.
For referring to `ClusterRoleBinding`, I mean `SA` will attached to its subject for sharing permission in related `ClusterRole`

```
$ kubectl get clusterrole federation-controller-manager:healthcheck-cluster1-cluster1 -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  creationTimestamp: 2019-03-01T08:40:23Z
  name: federation-controller-manager:healthcheck-cluster1-cluster1
  resourceVersion: "1100"
  selfLink: /apis/rbac.authorization.k8s.io/v1/clusterroles/federation-controller-manager%3Ahealthcheck-cluster1-cluster1
  uid: a7607cb7-3bfd-11e9-89b9-0800276f2b31
rules:
- nonResourceURLs:
  - /healthz
  verbs:
  - Get
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - list
$ kubectl get clusterrolebinding federation-controller-manager:healthcheck-cluster1-cluster1 -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  creationTimestamp: 2019-03-01T08:40:23Z
  name: federation-controller-manager:healthcheck-cluster1-cluster1
  resourceVersion: "1962"
  selfLink: /apis/rbac.authorization.k8s.io/v1/clusterrolebindings/federation-controller-manager%3Ahealthcheck-cluster1-cluster1
  uid: a762d03d-3bfd-11e9-89b9-0800276f2b31
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: federation-controller-manager:healthcheck-cluster1-cluster1
subjects:
- kind: ServiceAccount
  name: cluster1-cluster1
  namespace: foo
```

I will divide it to multiple PR and make generated code in independent commit.

**Comment by marun** [comment]
@xunpan Maybe I'm missing something, but why would it be desirable to share a RoleBinding vs just creating a new one for each SA?

**Comment by xunpan** [comment]
Close this ticket because I seperate the implementation in 4 PRs to make it more clear for review and more friendly history commit log.

The related PRs are:
#628 #629 #630 #631

**Pull request closed without merging**